packages/editor/src/lib/editor/derivations/spatialIndex.ts

```diff
@@ -0,0 +1,165 @@
+import { Editor } from '../Editor'
+
+type Element = {
+	minX: number
+	minY: number
+	maxX: number
+	maxY: number
+	id: TLShapeId
+}
+
+class TldrawRBush extends RBush<Element> {}
+
+export class SpatialIndex {
+	private readonly spatialIndex: ReturnType<typeof this.createSpatialIndex>
+	private lastPageId: TLPageId | null = null
+	private shapesInTree: Map<TLShapeId, Element>
+	private rBush: TldrawRBush
+
+	constructor(private editor: Editor) {
+		this.spatialIndex = this.createSpatialIndex()
+		this.shapesInTree = new Map<TLShapeId, Element>()
+		this.rBush = new TldrawRBush()
+	}
+
+	private addElement(id: TLShapeId, a: Element[], existingBounds?: Box) {
+		const e = this.getElement(id, existingBounds)
+		if (!e) return
+		a.push(e)
+		this.shapesInTree.set(id, e)
+	}
+
+	private getElement(id: TLShapeId, existingBounds?: Box): Element | null {
+		const bounds = existingBounds ?? this.editor.getShapeMaskedPageBounds(id)
+		if (!bounds) return null
+		return {
+			minX: bounds.minX,
+			minY: bounds.minY,
+			maxX: bounds.maxX,
+			maxY: bounds.maxY,
+			id,
+		}
+	}
+
+	private fromScratch(lastComputedEpoch: number) {
+		this.lastPageId = this.editor.getCurrentPageId()
+		this.shapesInTree = new Map<TLShapeId, Element>()
+		const elementsToAdd: Element[] = []
+
+		this.editor.getCurrentPageShapeIds().forEach((id) => {
+			this.addElement(id, elementsToAdd)
+		})
+
+		this.rBush = new TldrawRBush().load(elementsToAdd)
+
+		return lastComputedEpoch
+	}
+
+	private createSpatialIndex() {
+		const shapeHistory = this.editor.store.query.filterHistory('shape')
+
+		return computed<number>('spatialIndex', (prevValue, lastComputedEpoch) => {
+			if (isUninitialized(prevValue)) {
+				return this.fromScratch(lastComputedEpoch)
+			}
+
+			const diff = shapeHistory.getDiffSince(lastComputedEpoch)
+			if (diff === RESET_VALUE) {
+				return this.fromScratch(lastComputedEpoch)
+			}
+
+			const currentPageId = this.editor.getCurrentPageId()
+			if (!this.lastPageId || this.lastPageId !== currentPageId) {
+				return this.fromScratch(lastComputedEpoch)
+			}
+
+			let isDirty = false
+			for (const changes of diff) {
+				const elementsToAdd: Element[] = []
+				for (const id of Object.keys(changes.removed)) {
+					if (isShapeId(id)) {
+						const currentElement = this.shapesInTree.get(id)
+						if (currentElement) {
+							this.shapesInTree.delete(id)
+							this.rBush.remove(currentElement)
+							isDirty = true
+						}
+					}
+				}
+				this.shapesInTree.forEach((element, id) => {
+					const newBounds = this.editor.getShapeMaskedPageBounds(id)
+					if (!newBounds) return
+					if (
+						element.minX !== newBounds.minX ||
+						element.minY !== newBounds.minY ||
+						element.maxX !== newBounds.maxX ||
+						element.maxY !== newBounds.maxY
+					) {
+						this.shapesInTree.delete(id)
+						this.rBush.remove(element)
+						this.addElement(id, elementsToAdd, newBounds)
+						isDirty = true
+					}
+				})
```

Comment:
This is the main change. Instead of using the changes we go through all the existing shapes. 

We remove the `removed` first, so that we have fewer shapes to iterate through, and we only add `added` after this so we don't check those twice.

Reply:
Seems to be much faster to do it once per diff, even if there's only 1 change in diff 🤔